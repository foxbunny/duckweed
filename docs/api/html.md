# `duckweed.html()`

```javascript
(element, props, children) => VNode
```

The html function renders an element or a view function into a Snabbdom `VNode`
object. It's a thin wrapper around Snabbdom's `h()` function so you will find
references to Snabbdom documentation.

**WARNING:** If you pass a custom `patch()` function to the
[runner](./runner.md), the behavior described below may differ, or the function
may no longer work. You should use Snabbdom's `h()` instead in such cases.

The `html()` function accepts the following parameters.

---
| element | string or function | Either a HTML element name or a view function|
| props | object or null | The element attributes, event handler bindings, or view props|
| children | array of vnodes or a string | The nested elements, text nodes, and/or views|
---

## Element

Element can either be a function or a string. If you are not using JSX (or TSX)
you generally do not need to pass view functions to `html()`. You can invoke
them directly. The `html()` function was developed specifically for JSX so it
knows how to work with functions.

Here's an example of rendering a `DIV`:

```javascript
html("div");
```

## Props

Props are key-value pairs that have special meaning. You can use most HTML DOM
properties here, including `id`, `innerHTML`, `value`, and so on. For the most
part, they will work just like in HTML, but you should keep in mind that they
are actually DOM element properties, not HTML attributes.

```javascript
html("input", {value: "test", type: "password"});
```

There are some special props that are treated differently.

The `class` prop is special (note that DOM Elements don't have a `class`
property to begin with), and it manages the `className` property for you. The
behavior of this prop is a bit different from the [underlying Snabbdom
API](https://github.com/snabbdom/snabbdom#the-class-module). You can pass a
single class as a string, an array of class names as strings, or an object which
maps class names to their on/off state. Those of you coming from VueJS will feel
right at home. Let's look at the three options:

```javascript
// Just a string
html("div", {class: "menu"});

// Array of strings
html("div", {class: ["menu", "open"]});

// Object
html("div", {class: {menu: true, open: model.isOpen}});
```

The `on` prop is used to bind event handlers. It should be an object which maps
event names to the respective handlers. It is handled by the Snabbdom's
[eventlisteners
module](https://github.com/snabbdom/snabbdom#eventlisteners-module).

For convenience, `html()` also understands `on`-prefixed props, like `on-input`
or `on-click`, for convenience when using JSX:

```jsx
<div on-click={act("toggle")} />
```

The `hook` prop can be used to tap into different [Snabbdom
hooks](https://github.com/snabbdom/snabbdom#hooks). In general, you don't want
to abuse hooks too much. For example, patching the model inside an `update` hook
will cause an infinite loop. The Duckweed API does nothing to stop you from
shooting yourself in the foot when it comes to hooks.

The `style` prop can be used to specify element-specific styles and transitions.
It uses the Snabbdom's [style
module](https://github.com/snabbdom/snabbdom#the-style-module) under the hood.

You may also want to refer to [Non-standard
events](../guide/non-standard-events.md) guide for a few more of the props you
can use.

## Children

Children can either be a single string, or a mixed array of `VNode` objects,
`null`, and strings. They are rendered inside the element.

```javascript
html("div", {}, [
  html("span", "I'm a child"),
]);
```

When the element is a function, children are passed to it as a special `__inner`
prop. It is up to the function where it inserts them, if at all.

```jsx
const Aview = ({__inner}) => (
  <div class="parent">
    {__inner}
  </div>
);

const Bview = () => (
  <Aview>
    <span class="child">I'm a child</span>
  </Aview>
)
```

The `__inner` prop has a `__vnodes` property which contains an array of child
`VNode`, null, and/or string elements. This property is not generally meant to
be tinkered with, but you are free to do so if you know what you are doing.

[Documentation index](../main.md) | [API reference index](./main.md)
