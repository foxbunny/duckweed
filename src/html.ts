/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import * as snabbdom from "snabbdom";
import snab from "snabbdom/h";
import snabClass from "snabbdom/modules/class";
import snabEvents from "snabbdom/modules/eventlisteners";
import snabProps from "snabbdom/modules/props";
import styleModule from "snabbdom/modules/style";
import {VNode, VNodeData} from "snabbdom/vnode";

import documentevents from "./modules/documentevents";
import keyevents from "./modules/keyevents";
import routeevents from "./modules/routeevents";

type ChildVNodes = Array<VNode | null | undefined>;
type ChildVNodesArg = Array<VNode | undefined | null | InlineChild | [VNode] | string>;

interface InlineChild {
  __vnodes: ChildVNodes;
}

interface GenericProps {
  [prop: string]: any;
}

interface PropsBase {
  key?: string;
  prefix?: any[];
}

type ViewFunction<T = any> = (props?: T, children?: InlineChild) => VNode;

const patch = snabbdom.init([
  snabClass,
  styleModule,
  snabEvents,
  snabProps,
  documentevents,
  keyevents,
  routeevents,
]);

const isInlineChild = (obj: any): obj is InlineChild => {
  return typeof obj === "object" && obj !== null && typeof obj.vnodes !== "undefined";
};

const prepareClasses = (classes: string | string[] | {[name: string]: any} | null | undefined) => {
  if (classes == null) {
    return {};
  }
  if (typeof classes === "object" && !Array.isArray(classes)) {
    return classes;
  }
  if (typeof classes === "string") {
    return {[classes]: true};
  }
  return classes.reduce((o: {[name: string]: any}, c: string) => {
    o[c] = true;
    return o;
  }, {});
};

const prepareProps = (props: GenericProps | null): VNodeData => {
  if (props == null) {
    return {};
  }
  const finalProps: GenericProps = {};
  Object.keys(props).forEach((prop) => {
    const [mod, sub] = prop.split("-");
    if (sub) {
      finalProps[mod] = finalProps[mod] || {};
      finalProps[mod][sub] = props[prop];
    } else if (prop === "key") {
      finalProps.key = props[prop];
    } else if (prop === "on") {
      finalProps.on = props[prop];
    } else if (prop === "hook") {
      finalProps.hook = props[prop];
    } else if (prop === "class") {
      finalProps.class = prepareClasses(props[prop]);
    } else if (prop === "style") {
      finalProps.style = props[prop];
    } else if (prop === "route") {
      finalProps.route = props[prop];
    } else {
      finalProps.props = finalProps.props || {};
      finalProps.props[prop] = props[prop];
    }
  });
  return finalProps;
};

const renderIntrinsic = (elm: string, props: GenericProps = {}, children: ChildVNodesArg = []): VNode => {
  // FIXME: We're messing with any a lot here
  children = (children.length === 1
    ? children[0]
    : children.reduce((arr, c) => {
      if (isInlineChild(c)) {
        // Case where we have something like `{props.__inner}` somewhere in the
        // render functions.
        return arr.concat(c.__vnodes);
      }
      if (Array.isArray(c)) {
        // Case where we have something like `{arr.map(() => ...)}`
        return arr.concat(c);
      }
      return arr.concat([c as VNode]);
    }, [] as ChildVNodes)) as any;
  return snab(elm, prepareProps(props), children as any);
};

const renderFunction = (func: ViewFunction, props: any = {}, children: ChildVNodes = []): VNode => {
  const key = props && props.key;
  if (key) {
    delete props.key;
  }
  const vnode = func(props, {__vnodes: children || []});
  vnode.key = vnode.key || key;
  return vnode;
};

const html = (elm: string | ViewFunction, props?: any, ...children: ChildVNodesArg): VNode => {
  if (typeof elm === "string") {
    return renderIntrinsic(elm, props, children);
  } else {
    return renderFunction(elm, props, children as ChildVNodes);
  }
};

export {
  VNode,
  ChildVNodes,
  InlineChild,
  GenericProps,
  PropsBase,
  ViewFunction,
  patch,
  html,
};
export default html;
