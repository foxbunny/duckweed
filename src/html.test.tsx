/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import html from './html';

describe('html', () => {
  it('Should render a normal VNODE', () => {
    const vn = html('div');
    expect(vn).toMatchSnapshot();
  });

  it('Should accept normal DOM properties', () => {
    const vn = html('div', {title: 'undivided'});
    expect(vn).toMatchSnapshot();
  });

  it('Should pass keys', () => {
    const vn = html('div', {key: "don't change"});
    expect(vn).toMatchSnapshot();
  });

  it('Should pass event handlers', () => {
    const vn = html('div', {on: {click: jest.fn()}});
    expect(vn).toMatchSnapshot();
  });

  it('Should also accept shorthands for handlers', () => {
    const vn = html('div', {'on-click': jest.fn()});
    expect(vn).toMatchSnapshot();
  });

  it('Should accept hooks', () => {
    const vn = html('div', {hook: {init: jest.fn()}});
    expect(vn).toMatchSnapshot();
  });

  it('Should also accept shorthands for hooks', () => {
    const vn = html('div', {'hook-init': jest.fn()});
    expect(vn).toMatchSnapshot();
  });

  it('Should be able to render functions', () => {
    const vn = html(() => html('div'));
    expect(vn).toMatchSnapshot();
  });

  it('Should pass props to functions', () => {
    const v = ({foo}) => html('div', {}, foo);
    const vn = html(v, {foo: 'bar'});
    expect(vn).toMatchSnapshot();
  });

  it('Should pass a key to root vnode in a function', () => {
    const v = () => html('div');
    const vn = html(v, {key: "don't change"});
    expect(vn).toMatchSnapshot();
  });

  it('Should not freak out of props is null', () => {
    const vn = html('div', null);
    expect(vn).toMatchSnapshot();
    const vn2 = html(() => html('div'), null);
    expect(vn2).toMatchSnapshot();
  });

  it('Should render children', () => {
    const vn = html('div', {}, [
      html('span', {}, 'Hello'),
      html('span', {}, 'world'),
    ]);
    expect(vn).toMatchSnapshot();
  });
});

describe('html with jsx', () => {
  it('Should render JSX just fine', () => {
    const vn = (
      <div>Foo</div>
    );
    expect(vn).toMatchSnapshot();
  });

  it('Should handle children', () => {
    const vn = (
      <div><span>Foo</span></div>
    );
    expect(vn).toMatchSnapshot();
  });

  it('Should handle mixed text/vnode children', () => {
    const vn = (
      <p>
        go to
        <a href="#foo">foo</a>
      </p>
    );
    expect(vn).toMatchSnapshot();
  });

  it('Should handle arrays', () => {
    const a = [1, 2, 3];
    const vn = (
      <ul>
        {a.map(x => <li>{x}</li>)}
      </ul>
    );
    expect(vn).toMatchSnapshot();
  });

  it('should be able to add attributes to elements', () => {
    const vn = (
      <label _for="something">
        <input _id="something" />
      </label>
    );
    expect(vn).toMatchSnapshot();
  });
});
