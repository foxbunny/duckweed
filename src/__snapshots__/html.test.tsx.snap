// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html Should accept hooks 1`] = `
Object {
  "children": Array [],
  "data": Object {
    "hook": Object {
      "init": [Function],
    },
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should accept normal DOM properties 1`] = `
Object {
  "children": Array [],
  "data": Object {
    "props": Object {
      "title": "undivided",
    },
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should also accept shorthands for handlers 1`] = `
Object {
  "children": Array [],
  "data": Object {
    "on": Object {
      "click": [Function],
    },
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should also accept shorthands for hooks 1`] = `
Object {
  "children": Array [],
  "data": Object {
    "hook": Object {
      "init": [Function],
    },
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should be able to render functions 1`] = `
Object {
  "children": Array [],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should not freak out of props is null 1`] = `
Object {
  "children": Array [],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should not freak out of props is null 2`] = `
Object {
  "children": Array [],
  "data": Object {},
  "elm": undefined,
  "key": null,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should pass a key to root vnode in a function 1`] = `
Object {
  "children": Array [],
  "data": Object {},
  "elm": undefined,
  "key": "don't change",
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should pass event handlers 1`] = `
Object {
  "children": Array [],
  "data": Object {
    "on": Object {
      "click": [Function],
    },
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should pass keys 1`] = `
Object {
  "children": Array [],
  "data": Object {
    "key": "don't change",
  },
  "elm": undefined,
  "key": "don't change",
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should pass props to functions 1`] = `
Object {
  "children": undefined,
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": "bar",
}
`;

exports[`html Should render a normal VNODE 1`] = `
Object {
  "children": Array [],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html Should render children 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": Object {},
      "elm": undefined,
      "key": undefined,
      "sel": "span",
      "text": "Hello",
    },
    Object {
      "children": undefined,
      "data": Object {},
      "elm": undefined,
      "key": undefined,
      "sel": "span",
      "text": "world",
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html with jsx Should handle arrays 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": Object {},
      "elm": undefined,
      "key": undefined,
      "sel": "li",
      "text": 1,
    },
    Object {
      "children": undefined,
      "data": Object {},
      "elm": undefined,
      "key": undefined,
      "sel": "li",
      "text": 2,
    },
    Object {
      "children": undefined,
      "data": Object {},
      "elm": undefined,
      "key": undefined,
      "sel": "li",
      "text": 3,
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "ul",
  "text": undefined,
}
`;

exports[`html with jsx Should handle children 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": Object {},
      "elm": undefined,
      "key": undefined,
      "sel": "span",
      "text": "Foo",
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`html with jsx Should handle mixed text/vnode children 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": undefined,
      "elm": undefined,
      "key": undefined,
      "sel": undefined,
      "text": "go to",
    },
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "href": "#foo",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "a",
      "text": "foo",
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "p",
  "text": undefined,
}
`;

exports[`html with jsx Should render JSX just fine 1`] = `
Object {
  "children": undefined,
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": "Foo",
}
`;
